docs/0.19-changes.md

```diff
@@ -0,0 +1,95 @@
+
+```go
+validatorsProvider.Validators(ctx, "head", nil)
+```
+
+Even with a relatively simply call like this it is not clear what these parameters represent, especially when one is `nil`.  With the consolidation of call options this now looks something like this:
+
+```
+validatorsProvider.Validators(ctx, &api.ValidatorsOpts{
+  State: "head",
+})
+```
+
+Which makes it clearer that the `"head"` parameter refers to the state, and hides the explicit `nil` parameter for the specific validator indices.  Additional parameters added to the API can then be added to the struct without further disrupting the function signature.
+
+### Relocation of `Error` struct and extended usage
+`http.Error` has moved to `api.Error`, to make it clearer that it is reporting an error found within the API rather than specifically an HTTP error.
+
+Not all errors returned by the call will be `api.Error`, only those that return an error relating to the server.  So, for example, an attempt to call a function with a `nil` options struct (see below for options structures) will return a simple error.  But when an `api.Error` is available it will provide additional information about why the API rejected a request.
+
+### Response struct
+Responses from endpoints are now wrapped in the `api.Response` struct.  The requested information can be found in the `Data` element of this struct, and any metadata supplied in the response can be found in the `Metadata` element of this struct.  Metadata has been added to a number of endpoints to support situations where additional data may be useful to the user but is not part of the requested structure.  For example:
+
+```go
+response, _ := beaconBlockProvider.SignedBeaconBlock(ctx, &api.SignedBeaconBlockOpts{
+  Block: "123",
+})
+
+finalized, exists := response.Metadata[metadata.Finalized]
+if exists {
+  fmt.Printf("Is block finalized? %b\n",finalized)
+}
+```
+
+Note that different calls have different metadata; metadata keys are free-form text and information about which ones should be available for each call can be found in the beacon API documentation.
+
+### No more `nil, nil`
+Some functions could return `nil` for both result and error, which is bad practice.  This resulted in the following style of code being required:
+
+```go
+block, err := beaconBlockProvider.SignedBeaconBlock(ctx, "999999999999")
+if err != nil {
+  panic(err)
+}
+if block == nil {
+  panic("no block for this slot")
+}
+...
+```
+
+Functions will now either return a response or an error, but always one and never both.  Combined with use of `api.Error` as mentioned above this allows for better control of both successful and failed calls, for example:
+
+```go
+response, err := beaconBlockProvider.SignedBeaconBlock(ctx, &api.SignedBeaconBlockOpts{
+  Block: "999999999999",
+})
+if err != nil {
+var badInputErr *BadInputError
```

Comment:
this is not used in the code example below

Reply:
Fixed; thanks.